clap_complete/src/shells/shell.rs

```diff
@@ -36,10 +38,22 @@ impl FromStr for Shell {
     type Err = String;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let e = env::var("SHELL").ok().and_then(|x| {
+            PathBuf::from(x)
+                .file_name()
+                .and_then(|x| x.to_str())
+                .map(|x| x.to_string())
+        });
```

Comment:
A `FromStr` implementation should only be acting on the input, it shouldn't alternatively be parsing external state